File: 25.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        def reverse(node):
            head = node
            nodes_num = 1
            while(head.next):
                head = head.next
                nodes_num += 1
            pre_head = node
            if nodes_num < k:
                return head, pre_head, None
            next_node2 = node.next
            for _ in range(k-2):
                next_node = next_node2
                next_node2 = next_node.next
                next_node.next = node
                node = next_node
            next_head = next_node2.next
            next_node2.next = node
            return pre_head, next_node2, next_head

        if k == 1:
            return head
        phead = pre_head = ListNode(-1)
        # pre_head.next = head
        while head:
            end, head, next_head = reverse(head)
            pre_head.next = head
            pre_head = end
            pre_head.next = None
            # print(end, head, next_head)
            head = next_head
        return phead.next


if __name__ == "__main__":
    s = Solution()
    print()
    #第二次写了，希望再碰见能写出更简洁的

File: 25.go
package main

import "fmt"

/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseK(node *ListNode, k int) (head *ListNode, tail *ListNode, next *ListNode) {
	if node == nil {
		return
	}
	n := 0
	tmpNode := node
	for tmpNode != nil {
		n++
		tmpNode = tmpNode.Next
	}
	if n < k {
		return node, nil, nil
	}
	tail = node
	a, b := node, node.Next
	node.Next = nil //防止循环引用
	for i := 0; i < k-1; i++ {
		tmp := b.Next
		b.Next = a
		a, b = b, tmp
	}
	head = a
	next = b
	return
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{}
	preTail := dummy
	next := head
	var groupHead, groupTail *ListNode
	for next != nil {
		groupHead, groupTail, next = reverseK(next, k)
		preTail.Next = groupHead
		if groupTail == nil {
			return dummy.Next
		}
		preTail = groupTail
	}
	return dummy.Next
}

// @lc code=end

func main() {
	head := &ListNode{1, &ListNode{2, nil}}
	ans := reverseKGroup(head, 2)
	fmt.Println(ans.Val, ans.Next.Val)
}
